src/main/kotlin/io/emeraldpay/dshackle/upstream/AbstractHead.kt

```diff
@@ -85,9 +86,11 @@ abstract class AbstractHead @JvmOverloads constructor(
                     log.warn("Received signal $upstreamId $it unexpectedly - restart head")
                     lastHeadUpdated = 0L
                 } else {
-                    log.warn("Received signal $upstreamId $it - stop emit new head")
-                    completed = true
-                    stream.tryEmitComplete()
+                    if (completeHeadSub) {
+                        log.warn("Received signal $upstreamId $it - stop emit new head")
+                        completed = true
+                        stream.tryEmitComplete()
+                    }
```

Comment:
maybe we need log message on else branch also

Reply:
Agree